### DISCUSSION TITLE
Check WebSocket connection with the client

### DISCUSSION THREAD

**Discussion by rudmac** [fastapi/fastapi#8280] [Questions]
**How can I do a server heartbeats checks without send any information**

When I use websocket.application_state it returns CONNECTED, but when send some junk text within a 'disconnected' connection it raises an exception.

This is happening when I refresh my web page and the connection is lost, but the server doesn't know. I use only *server->client* message, one direction.

Is there any solution for that, I´m trying not use some send_text to do a heart beat.

Thanks,

**Comment by tiangolo** [comment]
WebSockets are a quite low-level set of APIs. In many cases, you have to implement the way you handle communication directly. So, in many cases, you might end up implementing REST stuff (HTTP) over WebSockets in your code.

And the implementations in clients are not necessarily as mature as we would wish. For example, there's no way to add headers to a WebSockets connection in browsers.

But given that, WebSockets have, as part of the spec, ping-pong definitions: https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#Pings_and_Pongs_The_Heartbeat_of_WebSockets

But you still have to send and receive them.

On the other side, you can simply catch the `starlette.websockets.Disconnect` exceptions.

**Comment by songzhi** [comment]
Hey Bro,I got the same problem,too.My current solution is do a custom ping-pong operation every several seconds.Here is my code:
```python
from starlette.websockets import WebSocket, WebSocketState
import asyncio

HEART_BEAT_INTERVAL = 5
async def is_websocket_active(ws: WebSocket) -> bool:
    if not (ws.application_state == WebSocketState.CONNECTED and ws.client_state == WebSocketState.CONNECTED):
        return False
    try:
        await asyncio.wait_for(ws.send_json({'type': 'ping'}), HEART_BEAT_INTERVAL)
        message = await asyncio.wait_for(ws.receive_json(), HEART_BEAT_INTERVAL)
        assert message['type'] == 'pong'
    except BaseException:  # asyncio.TimeoutError and ws.close()
        return False
    return True
```
If you have some tasks that will be awaited for long time,you should use `asyncio.wait_for`,too.Then catch the asyncio.TimeoutError,and `await is_websocket_active` again.If it is closed,you can quit the whole coroutine.There is also a important thing to be keep in mind,that is in your WebSocket client code,you should deal with the ping-pong message,too.
It works for me now,have any other idea?

**Comment by tiangolo** [comment]
Thanks for the note @songzhi !

@rudmac were you able to solve your problem?

**Comment by rudmac** [comment]
> Thanks for the note @songzhi !
> 
> @rudmac were you able to solve your problem?

Yes. I will close this issue. 
Thanks All

**Comment by mRokita** [comment]
I came up with the following solution:

```python
async for update in iter_updates():
    if not update:
        try:
            # Check if connection is active
            await asyncio.wait_for(websocket.receive_bytes(), timeout=1.0)
        except asyncio.TimeoutError:
            # Connection is still alive
            continue
        except WebSocketDisconnect:
            # Connection closed by client
            break
        else:
            # Received some data from the client, ignore it
            continue
    await websocket.send_text(update)
```

**Comment by msalbrain** [comment]
thanks a bunch @songzhi, I was trying to implement this but I wasn't that good with async progrms.

**Comment by yhgu2000** [comment]
I think the problem is clear: We lack support for the `ping/pong` mechanism specified in the [WebSocket standard](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#pings_and_pongs_the_heartbeat_of_websockets). As a reference, here is the solution of [the `websockets` library](https://websockets.readthedocs.io/en/stable/reference/asyncio/server.html#websockets.asyncio.server.ServerConnection.ping), which has two specific methods: `ping()` and `pong()` in addition to the `send()` and `recv()` methods. We just need something like that. I wonder if this is a design defect of `fastapi` or its underlying dependency `starlette`.

**Reply by yhgu2000** [reply]
If anyone could provide me with guidance, I would be willing to implement this feature.

**Comment by Raphasha27** [comment]
FastAPI's `application_state` only reflects the TCP-level state, not actual message reachability. A disconnected client won't send a close frame, so the server still sees `CONNECTED`.

To implement server-side heartbeat detection:

```python
from fastapi import FastAPI, WebSocket, WebSocketDisconnect
import asyncio

@app.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept()
    try:
        while True:
            # Send a ping frame (no data, just keepalive)
            await websocket.send_json({"type": "ping"})
            
            # Wait for pong response with timeout
            try:
                data = await asyncio.wait_for(
                    websocket.receive_json(), timeout=5.0
                )
                if data.get("type") == "pong":
                    continue  # client is alive
                # Handle other messages
                
            except asyncio.TimeoutError:
                # No response in 5 seconds — client disconnected
                print("Client heartbeat timed out, closing")
                break
                
    except WebSocketDisconnect:
        print("Client disconnected")
```

On the client side, respond to pings:

```javascript
ws.onmessage = (event) => {
    const data = JSON.parse(event.data);
    if (data.type === "ping") {
        ws.send(JSON.stringify({type: "pong"}));
    }
};
```

This way you detect silent disconnects (network drop, laptop sleep, etc.) within your timeout window, not relying on TCP state alone.
